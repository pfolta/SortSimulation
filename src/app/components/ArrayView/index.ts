export { default as ArrayView, ArrayViewProps } from "@/app/components/ArrayView/ArrayView";
export { default as Element, ElementProps, StyledElement } from "@/app/components/ArrayView/Element";
