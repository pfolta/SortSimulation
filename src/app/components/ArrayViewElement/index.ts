export type { ArrayViewElementProps } from "@/app/components/ArrayViewElement/ArrayViewElement";
export { default as ArrayViewElement } from "@/app/components/ArrayViewElement/ArrayViewElement";

export type { ImageElementProps } from "@/app/components/ArrayViewElement/ImageElement";
export { default as ImageElement, StyledImageElement } from "@/app/components/ArrayViewElement/ImageElement";

export type { NumberElementProps } from "@/app/components/ArrayViewElement/NumberElement";
export { default as NumberElement, StyledNumberElement } from "@/app/components/ArrayViewElement/NumberElement";
