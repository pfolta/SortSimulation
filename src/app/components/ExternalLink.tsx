import styled from "styled-components";

const ExternalLink = styled.a.attrs({
    target: "_blank",
    rel: "noreferrer"
})`
    :hover {
        text-decoration: none;
    }
`;

export default ExternalLink;
