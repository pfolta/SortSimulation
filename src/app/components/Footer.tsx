import React from "react";
import styled from "styled-components";

const StyledFooter = styled.footer`
    background-color: #e5e9f0;
    color: #4c566a;

    font-size: 0.875rem;

    padding: 1rem;
`;

const Footer = (): JSX.Element => <StyledFooter>Â© Peter Folta. All rights reserved.</StyledFooter>;

export default Footer;
