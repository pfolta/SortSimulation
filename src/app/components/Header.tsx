import React from "react";
import { HelpCircle, Moon, Settings, Sun } from "react-feather";
import styled from "styled-components";

import { useDispatch, useSelector } from "@/app/hooks";
import { toggleTheme } from "@/app/store";
import { themes } from "@/app/theme";

const StyledHeader = styled.header`
    background-color: peachpuff;

    display: flex;
    align-items: center;
    justify-content: space-between;

    position: sticky;
    top: 0;
`;

const Controls = styled.ul`
    display: flex;
`;

const Control = styled.li`
    list-style-type: none;
`;

const Button = styled.button`
    background: none;
    color: #2e3440;

    border: none;
    border-radius: 0.25rem;

    display: flex;
    padding: 0.25rem;

    :hover {
        background-color: rgba(0, 0, 0, 0.25);
    }
`;

const Header = (): JSX.Element => {
    const { theme } = useSelector((state) => state.settings);
    const dispatch = useDispatch();

    return (
        <StyledHeader>
            <h1>SortSimulation</h1>
            <Controls>
                <Control>
                    <Button onClick={() => dispatch(toggleTheme())}>{themes[theme] === themes.dark ? <Sun /> : <Moon />}</Button>
                </Control>
                <Control>
                    <Button>
                        <Settings />
                    </Button>
                </Control>
                <Control>
                    <Button>
                        <HelpCircle />
                    </Button>
                </Control>
            </Controls>
        </StyledHeader>
    );
};

export default Header;
