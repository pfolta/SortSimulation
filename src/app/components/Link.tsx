import styled from "styled-components";

const Link = styled.a`
    color: ${({ theme }) => theme.colors[theme.link.color]};
    cursor: pointer;

    :hover {
        text-decoration: none;
    }
`;

const ExternalLink = styled(Link).attrs({
    target: "_blank",
    rel: "noreferrer"
})``;

export { Link, ExternalLink };
