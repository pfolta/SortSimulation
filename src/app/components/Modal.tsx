import React, { ReactNode } from "react";
import { X } from "react-feather";
import { useIntl } from "react-intl";
import ReactModal, { Props as ReactModalProps } from "react-modal";
import styled from "styled-components";

import { BackgroundFadeIn, BackgroundFadeOut, SlideIn, SlideOut } from "@/app/animations";

interface ReactModalAdapterProps extends ReactModalProps {
    children?: ReactNode;
    className?: string;
}

const ReactModalAdapter = ({ className, ...delegated }: ReactModalAdapterProps) => (
    <ReactModal portalClassName={className} className={`${className}__Content`} overlayClassName={`${className}__Overlay`} {...delegated} />
);

const ANIMATION_DURATION = 250;

const StyledReactModal = styled(ReactModalAdapter).attrs({
    closeTimeoutMS: ANIMATION_DURATION
})`
    &__Overlay {
        display: grid;
        place-items: end center;

        position: fixed;
        top: 0;
        left: 0;
        width: 100vw;
        height: 100vh;
        z-index: 2;

        @media (min-width: ${({ theme }) => theme.modal.maxWidth}) {
            place-items: center;
        }

        &.ReactModal__Overlay--after-open {
            animation: ${({ theme }) => BackgroundFadeIn(theme.modal.background)} ${ANIMATION_DURATION}ms ease-out forwards;
        }

        &.ReactModal__Overlay--before-close {
            animation: ${({ theme }) => BackgroundFadeOut(theme.modal.background)} ${ANIMATION_DURATION}ms ease-in forwards;
        }
    }

    &__Content {
        display: grid;
        position: relative;

        background: ${({ theme }) => theme.body.background};
        box-shadow: 0 0.5rem 2rem 0 rgba(0, 0, 0, 0.75);
        border-radius: 0.5rem 0.5rem 0 0;

        width: min(${({ theme }) => theme.modal.maxWidth}, 100vw);
        max-height: calc(100vh - ${({ theme }) => theme.modal.marginTop});

        outline: none;
        overflow: hidden;

        @media (min-width: ${({ theme }) => theme.modal.maxWidth}) {
            border-radius: 0.5rem;
        }

        & > * {
            padding: ${({ theme }) => theme.modal.padding};
            overflow: auto;
            max-height: calc(100vh - ${({ theme }) => theme.modal.marginTop} - 2 * ${({ theme }) => theme.modal.padding});
        }

        &.ReactModal__Content--after-open {
            animation: ${SlideIn} ${ANIMATION_DURATION}ms ease-out forwards;
        }

        &.ReactModal__Content--before-close {
            animation: ${SlideOut} ${ANIMATION_DURATION}ms ease-in forwards;
        }
    }
`;

const CloseButton = styled.button`
    all: unset;

    display: grid;
    padding: 0.25rem;

    position: absolute;
    top: 1rem;
    right: 1rem;

    background-color: ${({ theme }) => theme.colors.gray5};
    color: ${({ theme }) => theme.colors.gray3};

    border-radius: 0.25rem;

    :focus {
        box-shadow: ${({ theme }) => theme.focus.boxShadow};
    }

    :hover {
        color: ${({ theme }) => theme.colors.gray1};
    }

    :active {
        background-color: ${({ theme }) => theme.colors.gray4};
    }
`;

interface ModalProps {
    children?: ReactNode;
    isOpen: boolean;
    onClose: () => void;
}

const Modal = ({ children, onClose, isOpen }: ModalProps): JSX.Element => {
    const intl = useIntl();

    const disableBodyScroll = () => (document.body.style.overflow = "hidden");
    const enableBodyScroll = () => (document.body.style.overflow = "unset");

    return (
        <StyledReactModal isOpen={isOpen} onRequestClose={onClose} onAfterOpen={disableBodyScroll} onAfterClose={enableBodyScroll}>
            <CloseButton title={intl.formatMessage({ id: "close" })} onClick={onClose}>
                <X size="1.25rem" />
            </CloseButton>
            {children}
        </StyledReactModal>
    );
};

export default Modal;
