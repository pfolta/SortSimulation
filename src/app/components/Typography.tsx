import styled from "styled-components";

const Title = styled.h1`
    font-size: 2rem;
    font-weight: bold;
`;

const Heading = styled.h2`
    font-size: 1.5rem;
    font-weight: bold;
    margin-bottom: 1rem;
`;

const SubHeading = styled.h3`
    color: ${({ theme }) => theme.colors[theme.typography.subheading.color]};
    font-weight: bold;
    text-transform: uppercase;
`;

const Paragraph = styled.p`
    margin-bottom: 1rem;

    :last-child {
        margin-bottom: 0;
    }
`;

export { Title, Heading, SubHeading, Paragraph };
