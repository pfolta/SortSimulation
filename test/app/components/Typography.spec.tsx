import React from "react";

import renderWithProviders from "../../utils/renderWithProviders";

import { Title, Heading, SubHeading, Paragraph } from "@/app/components/Typography";

describe("Typography", () => {
    describe("Title", () => {
        it("renders correctly", () => {
            expect(renderWithProviders(<Title>Title</Title>).toJSON()).toMatchSnapshot();
        });
    });

    describe("Heading", () => {
        it("renders correctly", () => {
            expect(renderWithProviders(<Heading>Heading</Heading>).toJSON()).toMatchSnapshot();
        });
    });

    describe("SubHeading", () => {
        it("renders correctly", () => {
            expect(renderWithProviders(<SubHeading>SubHeading</SubHeading>).toJSON()).toMatchSnapshot();
        });
    });

    describe("Paragraph", () => {
        it("renders correctly", () => {
            expect(renderWithProviders(<Paragraph>Paragraph</Paragraph>).toJSON()).toMatchSnapshot();
        });
    });
});
