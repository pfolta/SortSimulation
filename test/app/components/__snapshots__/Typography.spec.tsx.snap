// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography Heading renders correctly 1`] = `
.c0 {
  font-size: 1.5rem;
  font-weight: bold;
  margin-bottom: 1rem;
}

<h2
  className="c0"
>
  Heading
</h2>
`;

exports[`Typography Paragraph renders correctly 1`] = `
.c0 {
  margin-bottom: 1rem;
}

.c0:last-child {
  margin-bottom: 0;
}

<p
  className="c0"
>
  Paragraph
</p>
`;

exports[`Typography SubHeading renders correctly 1`] = `
.c0 {
  color: #666666;
  font-weight: bold;
  text-transform: uppercase;
}

<h3
  className="c0"
>
  SubHeading
</h3>
`;

exports[`Typography Title renders correctly 1`] = `
.c0 {
  font-size: 2rem;
  font-weight: bold;
}

<h1
  className="c0"
>
  Title
</h1>
`;
